modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportReindexAction.java

```diff
@@ -0,0 +1,273 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.reindex;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.bulk.BulkItemResponse.Failure;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.action.search.ShardSearchFailure;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.AutoCreateIndex;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.cluster.ClusterService;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.lucene.uid.Versions;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.index.mapper.internal.TTLFieldMapper;
+import org.elasticsearch.index.mapper.internal.VersionFieldMapper;
+import org.elasticsearch.script.ScriptService;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.List;
+import java.util.Objects;
+
+import static java.util.Objects.requireNonNull;
+import static org.elasticsearch.index.VersionType.INTERNAL;
+
+public class TransportReindexAction extends HandledTransportAction<ReindexRequest, ReindexResponse> {
+    private final ClusterService clusterService;
+    private final ScriptService scriptService;
+    private final AutoCreateIndex autoCreateIndex;
+    private final Client client;
+
+    @Inject
+    public TransportReindexAction(Settings settings, ThreadPool threadPool, ActionFilters actionFilters,
+            IndexNameExpressionResolver indexNameExpressionResolver, ClusterService clusterService, ScriptService scriptService,
+            AutoCreateIndex autoCreateIndex, Client client, TransportService transportService) {
+        super(settings, ReindexAction.NAME, threadPool, transportService, actionFilters, indexNameExpressionResolver,
+                ReindexRequest.class);
+        this.clusterService = clusterService;
+        this.scriptService = scriptService;
+        this.autoCreateIndex = autoCreateIndex;
+        this.client = client;
+    }
+
+    @Override
+    protected void doExecute(Task task, ReindexRequest request, ActionListener<ReindexResponse> listener) {
+        validateAgainstAliases(request.getSearchRequest(), request.getDestination(), indexNameExpressionResolver, autoCreateIndex,
```

Comment:
@imotov had the same idea. Maybe it is a thing we should think about in
5.x.

Oh!  I left a comment earlier about resolving issues but I just forgot
resolve the "don't run this on a mixed cluster" thing. I'll get that in the
morning.
On Mar 13, 2016 5:11 PM, "Boaz Leskes" <notifications@github.com> wrote:

> In
> modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportReindexAction.java
> <https://github.com/elastic/elasticsearch/pull/17060#discussion_r55943844>
> :
>
> > +
> > +    @Inject
> > +    public TransportReindexAction(Settings settings, ThreadPool threadPool, ActionFilters actionFilters,
> > +            IndexNameExpressionResolver indexNameExpressionResolver, ClusterService clusterService, ScriptService scriptService,
> > +            AutoCreateIndex autoCreateIndex, Client client, TransportService transportService) {
> > +        super(settings, ReindexAction.NAME, threadPool, transportService, actionFilters, indexNameExpressionResolver,
> > +                ReindexRequest.class);
> > +        this.clusterService = clusterService;
> > +        this.scriptService = scriptService;
> > +        this.autoCreateIndex = autoCreateIndex;
> > +        this.client = client;
> > +    }
> > +
> > +    @Override
> > +    protected void doExecute(Task task, ReindexRequest request, ActionListener<ReindexResponse> listener) {
> > +        validateAgainstAliases(request.getSearchRequest(), request.getDestination(), indexNameExpressionResolver, autoCreateIndex,
>
> I wondered in the past whether we should add something to the transport
> registration logic indicating which action were added on what version, so
> we can automatically disable (new) actions if any of the nodes in the
> cluster is too old. Feels like this kind of things belong in a generic
> infra so action implementers won't need to worry about them...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/17060/files#r55943844>.
>